charms/unit/harness.py

```diff
@@ -0,0 +1,122 @@
+#
+# harness.py -- Unit Testing Harness for Layer Charms
+#
+# copyright 2016 Canonical Ltd.
+# TODO: License (Apache or LGPL?)
+#
+
+from glob import glob
+import mock
+import unittest
+
+class HarnessError(Exception): pass
+
+class Harness(unittest.TestCase):
+    '''
+    I am a unit testing harness for Juju Layered charms.
+
+    I automatically mock out any modules referenced in the layers
+    directory that do not actually exist in the layer's source. I
+    maintain an internal table of the mock objects, so that you may
+    add behavior to them if necessary for a test.
+
+    I will also setup a mock for the status handlers in hookenv, so
+    that you may set and confirm status in a test.
+
+    '''
+    def __init__(self, *args, **kwargs):
+        super(Harness, self).__init__(*args, **kwargs)
+        self._to_mock = []
+        self._patchers = []
+        self._local_modules = None
+        self.mocks = {}
+
+    @property
+    def last_status(self):
+        '''
+        Helper for mocked out status list.
+
+        Returns the last status set. Raises a HarnessError if this
+        method is called, but the hookenv mock is not active.
+
+        '''
+        try:
+            return self.statuses[-1]
+        except AttributeError:
+            raise HarnessError(
+                'self.statuses does not exist. Did you initialize a '
+                'harness without passing in mock_hookenv_status=True?')
+
+    @property
+    def local_modules(self):
+        '''
+        Searches the source tree for Python files. Returns a list of
+        module names, each suitable for passing to mock.patch.
+
+        This method will only run once per class instantiation.
+
+        '''
+        if self._local_modules is None:
+            # TODO: there is probably a more elegant way to do
+            # this. At the very least, we probably want to handle
+            # Python files without a .py extension.
+
+            # Grab everything in reactive and lib/charms/layer
+            mods = glob("**reactive/*.py", recursive=True)
+            mods.extend(glob("**lib/charms/layer/*.py", recursive=True))
```

Comment:
The recursive support was added in 3.5 and won't work on trusty (which has 3.4).